File: ascii_fields/animations/whirlpool.py
import math

from ..core import Animation, clamp, render_field
from ..noise import fbm


class WhirlpoolAnimation(Animation):
  """A maelstrom: water spiralling inward and down a dark central throat, with
  logarithmic spiral arms, foam streaks and a churning surface."""

  thresholds = [
    (0.10, " "), (0.20, "."), (0.32, ":"), (0.44, "-"), (0.56, "="),
    (0.68, "+"), (0.80, "*"), (0.90, "#"), (1.01, "@"),
  ]

  def render(self, width, height, elapsed, phase, options):
    t = elapsed
    arms = 2
    windings = 5.0 * max(0.5, options.scale)
    contrast = options.contrast
    radius = max(1.0, min(width, height * 2) * 0.5)
    grid = []
    for row in range(height):
      py = ((row - (height - 1) * 0.5) * 2.0) / radius
      line = []
      for col in range(width):
        px = (col - (width - 1) * 0.5) / radius
        r = math.hypot(px, py) + 1e-4
        theta = math.atan2(py, px)
        # faster rotation toward the centre (differential swirl)
        swirl = theta + windings * math.log(r + 0.05) + t * (0.6 + 1.4 / (1.0 + 8.0 * r))
        spiral = 0.5 + 0.5 * math.sin(arms * swirl)
        foam = fbm(math.cos(swirl) * 3.0 + t * 0.2, math.sin(swirl) * 3.0, octaves=4)
        surface = 0.45 * spiral + 0.45 * foam
        throat = 1.0 - math.exp(-(r * r) / 0.02)       # dark hole in the middle
        rim = math.exp(-((r - 0.16) ** 2) / 0.01) * 0.6
        edge = clamp(1.0 - (r - 1.0) * 2.5)            # fade past the rim
        level = (surface * throat + rim) * edge
        line.append(clamp(level * 1.5 * contrast))
      grid.append(line)
    return render_field(width, height, grid, options, self)

File: src/animations/whirlpool.rs
use rayon::prelude::*;

use crate::animation::{Animation, FrameContext};
use crate::core::{clamp, render_field, FieldStyle};
use crate::noise::fbm;

const STYLE: FieldStyle = FieldStyle { gray_lo: 234, gray_hi: 255, default_theme: "ocean" };
const TH: &[(f64, char)] = &[
  (0.10, ' '), (0.20, '.'), (0.32, ':'), (0.44, '-'), (0.56, '='),
  (0.68, '+'), (0.80, '*'), (0.90, '#'), (1.01, '@'),
];

pub struct Whirlpool;

impl Animation for Whirlpool {
  fn render(&mut self, ctx: &FrameContext, out: &mut String) {
    let w = ctx.width; let h = ctx.height;
    let mut grid = vec![0.0_f64; w * h];
    let t = ctx.elapsed;
    let arms = 2.0_f64;
    let windings = 5.0 * ctx.options.scale.max(0.5);
    let contrast = ctx.options.contrast;
    let radius = (w.min(h * 2) as f64 * 0.5).max(1.0);
    let cx_mid = (w as f64 - 1.0) * 0.5;
    let cy_mid = (h as f64 - 1.0) * 0.5;
    grid.par_chunks_mut(w).enumerate().for_each(|(row, row_slice)| {
      let py = ((row as f64 - cy_mid) * 2.0) / radius;
      for col in 0..w {
        let px = (col as f64 - cx_mid) / radius;
        let r = (px * px + py * py).sqrt() + 1e-4;
        let theta = py.atan2(px);
        let swirl = theta + windings * (r + 0.05).ln() + t * (0.6 + 1.4 / (1.0 + 8.0 * r));
        let spiral = 0.5 + 0.5 * (arms * swirl).sin();
        let foam = fbm(swirl.cos() * 3.0 + t * 0.2, swirl.sin() * 3.0, 4);
        let surface = 0.45 * spiral + 0.45 * foam;
        let throat = 1.0 - (-(r * r) / 0.02).exp();
        let rim = (-((r - 0.16).powi(2)) / 0.01).exp() * 0.6;
        let edge = (1.0 - (r - 1.0) * 2.5).max(0.0).min(1.0);
        let level = (surface * throat + rim) * edge;
        row_slice[col] = clamp(level * 1.5 * contrast);
      }
    });
    render_field(ctx, &grid, TH, &STYLE, out);
  }
}
